html/controls/operator.js

```diff
@@ -1089,6 +1088,212 @@ function createJamDialog() {
 	});
 }
 
+function createTimeoutDialog() {
+	var lastPeriodRegistered = 0;
+	var lastPeriodListed = 0;
+	var firstJamListed = [0];
+	var lastJamListed = [0];
+	var periodDropdownTemplate = $('<select>').attr('id', 'PeriodDropdown')
+		.append($('<option>').attr('value', 0).text('P0'));
+	var jamDropdownTemplate = [$('<select>').attr('id', 'JamDropdown').attr('period', 0)
+		.append($('<option>').attr('value', 0).text('J0'))];
+	var typeDropdownTemplate = $('<select>').attr('id', 'TypeDropdown')
+		.append($('<option>').attr('value', '.false').text('No type'))
+		.append($('<option>').attr('value', 'O.false').text('Off. Timeout'))
+		.append($('<option>').attr('value', '1.false').text('Team TO left'))
+		.append($('<option>').attr('value', '2.false').text('Team TO right'))
+		.append($('<option>').attr('value', '1.true').text('Off. Review left'))
+		.append($('<option>').attr('value', '2.true').text('Off. Review right'));
+	
+	var dialog = $("<div>").addClass("NumberDialog");
+	var table = $("<table>").appendTo(dialog);
+	var headers = $("<tr><td/><td/><td/><td/><td/><td/></tr>").appendTo(table);
+	$("<a>").text("Period").addClass("Title")
+		.appendTo(headers.find("td:eq(0)").addClass("Title"));
+	$("<a>").text("After Jam").addClass("Title")
+		.appendTo(headers.find("td:eq(1)").addClass("Title"));
+	$("<a>").text("Duration").addClass("Title")
+		.appendTo(headers.find("td:eq(2)").addClass("Title"));
+	$("<a>").text("Type").addClass("Title")
+		.appendTo(headers.find("td:eq(3)").addClass("Title"));
+	
+	var footer = $("<tr><td/><td/><td/><td/><td/><td/></tr>").attr('id', 'toFooter').appendTo(table);
+	periodDropdownTemplate.clone().appendTo(footer.find('td:eq(0)'));
+	$('<button>').text('Add Timeout').button().click(function() {
+		WS.Set('ScoreBoard.Period('+footer.find('#PeriodDropdown').val()+').InsertTimeout', true);
+	}).appendTo(footer.find('td:eq(3)'));
+	
+	WS.Register(['ScoreBoard.CurrentPeriodNumber'], function(k, v) {
+		var p = Number(v);
+		var i;
+		for (i = lastPeriodRegistered + 1; i <= p; i++) {
+			firstJamListed[p] = 0;
+			lastJamListed[p] = 0;
+			jamDropdownTemplate[i] = $('<select>').attr('id', 'JamDropdown').attr('period', i);
+			WS.Register(['ScoreBoard.Period('+i+').FirstJamNumber'], processFirstJamNumber);
```

Comment:
Use one register for all of this, it's slow and leaks otherwise.

Reply:
Using one Register would require registering `ScoreBoard.Period` I'm pretty certain that's slower just because of the huge amount of events that would feed into the method.

And it's not very leaky as I'm only registering once for each period number and it's safe to assume we're not going to see a lot of different period numbers over the lifetime of the screen.

Reply:
We already register all of ScoreBoard.Period, and over a tournament it could potentially be 50+ periods.

Reply:
We don't call the methods used here for all of ScoreBoard.Period.
And amongst those 50+ Periods there will be none with a number greater than 2. So we register our 3 methods once each for ScoreBoard.Period(0).X, ScoreBoard.Period(1).X, and ScoreBoard.Period(2).X That's 9 registrations. I'm certain that they will not be more costly than tens of thousands of method invocations.

Reply:
That's 9 registrations, per game. There can be more than one game, and each of those will then leak.

I've put quite a bit of effort into optimising all the new stats pages, and removing registrations such as this are a big part of it. Please keep to solely top-level registrations.

Reply:
I DO NOT reregister for each game. I register each period NUMBER once over the lifetime of the screen. That's why I distinguish between `lastPeriodRegistered` and `lastPeriodListed`.

THe other items are in now.

Reply:
Do you have benchmarks indicating this makes a notable difference? I'd prefer to keep a consistent message about how to do things efficiently.

Reply:
I don't have benchmarks. I also don't see an inconsistent message. The rule is "Only use IDs in your calls to WS.Register, if the number of different IDs encountered over the lifetime of the screen remains small." (There are tons of calls with `Team(id)` in them, including in your optimised modules. Which is fine, as we'll only ever see two team ids.)

Reply:
It's more nuanced than that, it matters also how many times you call it - calling it multiple times with different context is just as bad. I'd rather others not copy&paste and cause the problems I just finished fixing.

Reply:
That's what we review code for (amongst other reasons).

Reply:
Yes, and I'm pointing out the performance issues here. Anything not top-level will slow down page load, as it will trigger a registration to the backend, and then every registered function that matches it - so that's multiple RTTs at startup and an extra few DOM paintings. Please change this, this pattern needs to be eliminated from the codebase.